src/helpers/pull-helpers/pull-request-parsing.ts

```diff
@@ -0,0 +1,104 @@
+import { encode } from "gpt-tokenizer";
+import { TokenLimits } from "../../types/llm";
+import { logger } from "../errors";
+import { EncodeOptions } from "gpt-tokenizer/esm/GptEncoding";
+
+export async function processPullRequestDiff(diff: string, tokenLimits: TokenLimits) {
+  const { runningTokenCount, tokensRemaining } = tokenLimits;
+
+  // parse the diff into per-file diffs for quicker processing
+  const perFileDiffs = parsePerFileDiffs(diff);
+
+  // quick estimate using a simple heuristic; 3.5 characters per token
+  const estimatedFileDiffStats = perFileDiffs.map(({ filename, diffContent }) => {
+    const estimatedTokenCount = Math.ceil(diffContent.length / 3.5);
+    return { filename, estimatedTokenCount, diffContent };
+  });
+
+  estimatedFileDiffStats.sort((a, b) => a.estimatedTokenCount - b.estimatedTokenCount); // Smallest first
+
+  let currentTokenCount = runningTokenCount;
+  const includedFileDiffs = [];
+
+  // Using the quick estimate, include as many files as possible without exceeding token limits
+  for (const file of estimatedFileDiffStats) {
+    if (currentTokenCount + file.estimatedTokenCount > tokensRemaining) {
+      logger.info(`Skipping ${file.filename} to stay within token limits.`);
+      continue;
+    }
+    includedFileDiffs.push(file);
+    currentTokenCount += file.estimatedTokenCount;
+  }
+
+  // If no files can be included, return null
+  if (includedFileDiffs.length === 0) {
+    logger.error(`Cannot include any files from diff without exceeding token limits.`);
+    return { diff: null };
+  }
+
+  // Accurately calculate token count for included files we have approximated to be under the limit
+  const accurateFileDiffStats = await Promise.all(
+    includedFileDiffs.map(async (file) => {
+      const tokenCountArray = await encodeAsync(file.diffContent, { disallowedSpecial: new Set() });
+      const tokenCount = tokenCountArray.length;
+      return { ...file, tokenCount };
+    })
+  );
+
+  // Take an accurate reading of our current collection of files within the diff
+  currentTokenCount = accurateFileDiffStats.reduce((sum, file) => sum + file.tokenCount, runningTokenCount);
+
+  // Remove files from the end of the list until we are within token limits
+  while (currentTokenCount > tokensRemaining && accurateFileDiffStats.length > 0) {
+    const removedFile = accurateFileDiffStats.pop();
+    currentTokenCount -= removedFile?.tokenCount || 0;
+    logger.info(`Excluded ${removedFile?.filename || "Unknown filename"} after accurate token count exceeded limits.`);
+  }
```

Comment:
I think we should have a list of files (maybe configurable) that we will always remove like `bun.lockb`

Reply:
For now can you give the names of files that should be removed?

Reply:
`yarn.lock`, `pnpm-lock.yaml`, `bun.lockb`, `package-lock.json`

Reply:
A better approach would be to import this from some type of maintained library. There's so many more of these lock files in other programming language ecosystems. 

Hard coding a list like this is almost always not the best approach. 

We can also follow some heuristics like never attempting to parse binary files which I presume is buns lock file format.

Reply:
It's almost impossible to cover all cases, that's why I suggested that we add it to the config so it can be configurable if needed, however I realize that it's not ideal as we want things to be plug-and-play.